docs/proposals/protobuf.md

```diff
@@ -0,0 +1,501 @@
+is expected to be of type `watch.Event`, and is currently not self describing.
+
+For expediency and consistency, we define a format for Protobuf watch Events
+that is similar. Since protobuf messages are not self describing, we must
+identify the boundaries between Events (a `frame`). We do that by prefixing
+each frame of N bytes with a 4-byte, big-endian, unsigned integer with the
+value N.
+
+    frame  = length body
+    length = 32-bit unsigned integer in big-endian order, denoting length of
+             bytes of body
+    body = <bytes>
+
+    # frame containing a single byte 0a
+    frame = 01 00 00 00 0a
+
+    # equivalent JSON
+    frame = {"type": "added", ...}
+
+The body of each frame is a serialized Protobuf message `Event` in package
+`k8s.io/kubernetes/pkg/watch/versioned`. The content type used for this
+format is `application/vnd.kubernetes.protobuf;type=watch`.
+
+## Negotiation
+
+To allow clients to request protobuf serialization optionally, the `Accept`
+HTTP header is used by callers to indicate which serialization they wish
+returned in the response, and the `Content-Type` header is used to tell the
+server how to decode the bytes sent in the request (for DELETE/POST/PUT/PATCH
+requests). The server will return 406 if the `Accept` header is not
+recognized or 415 if the `Content-Type` is not recognized (as defined in
+RFC2616).
+
+To be backwards compatible, clients must consider that the server does not
+support protobuf serialization. A number of options are possible:
+
+### Preconfigured
+
+Clients can have a configuration setting that instructs them which version
+to use. This is the simplest option, but requires intervention when the
+component upgrades to protobuf.
+
+### Include serialization information in api-discovery
+
+Servers can define the list of content types they accept and return in
+their API discovery docs, and clients can use protobuf if they support it.
+Allows dynamic configuration during upgrade if the client is already using
+API-discovery.
+
+### Optimistically attempt to send and receive requests using protobuf
+
+Using multiple `Accept` values:
+
+    Accept: application/vnd.kubernetes.protobuf, application/json
+
+clients can indicate their preferences and handle the returned
+`Content-Type` using whatever the server responds. On update operations,
+clients can try protobuf and if they receive a 415 error, record that and
+fall back to JSON. Allows the client to be backwards compatible with
+any server, but comes at the cost of some implementation complexity.
+
+
+## Generation process
+
+Generation proceeds in five phases:
+
+1. Generate a gogo-protobuf annotated IDL from the source Go struct.
+2. Generate temporary Go structs from the IDL using gogo-protobuf.
+3. Generate marshaller/unmarshallers based on the IDL using gogo-protobuf.
+4. Take all tag numbers generated for the IDL and apply them as struct tags
+   to the original Go types.
```

Comment:
It does not require it (we could avoid generating the tags) but I think we
want people to be able to reorder the gostructs if they want internally.

If you add a new type, when you run generation the first time your protobuf
IDL and generated serializers will be created, AND the source type fields
will have protobuf struct tags added.  The verify-generated-protobuf script
should be checking to verify that those didn't change (and it would fail,
telling you that there is a different set of code).

On Mon, Mar 7, 2016 at 4:24 AM, Wojciech Tyczynski <notifications@github.com
> wrote:

> In docs/proposals/protobuf.md
> <https://github.com/kubernetes/kubernetes/pull/22600#discussion_r55184663>
> :
>
> > +clients can indicate their preferences and handle the returned
> > +`Content-Type` using whatever the server responds. On update operations,
> > +clients can try protobuf and if they receive a 415 error, record that and
> > +fall back to JSON. Allows the client to be backwards compatible with
> > +any server, but comes at the cost of some implementation complexity.
> > +
> > +
> > +## Generation process
> > +
> > +Generation proceeds in five phases:
> > +
> > +1. Generate a gogo-protobuf annotated IDL from the source Go struct.
> > +2. Generate temporary Go structs from the IDL using gogo-protobuf.
> > +3. Generate marshaller/unmarshallers based on the IDL using gogo-protobuf.
> > +4. Take all tag numbers generated for the IDL and apply them as struct tags
> > +   to the original Go types.
>
> So IIUC, that means that supporting protobuffers will require additional
> tags in the original files (this will be a corresponding tag to currently
> existing "json" tag). Am I right?
>
> If I understood correctly, if you decide to add a new type to an API, then
> it will be "corrected" by generators, by adding appropriate tag numbers as
> tags, right? This is nice, but I think it won't be obvious for people and
> we probably need to clearly document it.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/22600/files#r55184663>.
>